Add a LowerInt64 B3 pass

aoikonomopoulos opened the pull request:
#### e4b1106bc45645cdf986bb2b174622b82c8b03bd
<pre>
Add a LowerInt64 B3 pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=275988">https://bugs.webkit.org/show_bug.cgi?id=275988</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

Base infrastructure and partial implementation of a lowering pass for
Int64 values on 32-bits.

Crucially, this doesn&apos;t lower Patchpoints/CCalls and Add/Sub on Int64
values.  The former would, at the moment, complicate the generic paths
that simultaneously iterate over B3 Values and Air args. The latter
(Add/Sub) is lowered to the pre-existing Add64/Sub64 Air forms, which
avoids the need to track the carry bit.

Since we need to interface with the above values that use Int64s, this
patch also introduces two new B3 opcodes (that are only used on
32-bits):

- TruncHigh: extracts the high Int32 of an Int64 value.
- Stitch: creates an Int64 out of two Int32 values.

They need to be defined on 64-bits too because B3ValueInlines.h
references opcodes in a macro definition.

This removes the previously added lowering code to
B3LowerToAir32_64.cpp. It implements enough ops in B3LowerInt64.cpp to
keep i64-call.js and cc-int-to-int.js working with OMG on ARMv7.

It also fixes a bug in the Air forms for Add64/Sub64.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3Generate.cpp:
(JSC::B3::generateToAir):
* Source/JavaScriptCore/b3/B3LowerInt64.cpp: Added.
(JSC::B3::lowerInt64):
* Source/JavaScriptCore/b3/B3LowerInt64.h: Added.
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3StackmapSpecial.cpp:
(JSC::B3::StackmapSpecial::isArgValidForRep):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
(JSC::B3::Value::typeFor):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueRep.h:
(JSC::B3::ValueRep::ValueRep):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:

Canonical link: <a href="https://commits.webkit.org/280764@main">https://commits.webkit.org/280764@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b971006fbf5341a164bbc5a1c508e4ce73da6c48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7911 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8099 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46537 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5604 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49638 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27401 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31306 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6941 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6914 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50558 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53283 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62767 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56708 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7306 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53796 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53891 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12726 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1179 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78469 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32623 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13016 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33708 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34793 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3ValueRep.h:
```diff
@@ -129,7 +129,22 @@ class ValueRep {
+            || kind == SomeEarlyRegister
+            || kind == SomeLateRegister
+#if USE(JSVALUE32_64)
+            || kind == SomeRegisterPair
```
Do we really need this many variants?

aoikonomopoulos replied:
Not at the moment, no. I think it's easier to remove those we don't end up using than adding individual variants as they are needed though.

FWIW the variants are already there, it's just that this assert had not been updated to account for them. If we want to remove them and add them back as needed, I'd submit a separate PR for that.

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerInt64.cpp:
```diff
@@ -0,0 +1,701 @@
+
+namespace B3LowerInt64Internal {
+static constexpr bool verbose = false;
+static constexpr bool selfTest = false;
```
this is unused, right?

aoikonomopoulos replied:
Yah, it's a leftover from an earlier iteration of the PR, I'll GC it.

justinmichaud reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

aoikonomopoulos committed:
Add a LowerInt64 B3 pass
https://bugs.webkit.org/show_bug.cgi?id=275988

Reviewed by Yusuke Suzuki and Justin Michaud.

Base infrastructure and partial implementation of a lowering pass for
Int64 values on 32-bits.

Crucially, this doesn't lower Patchpoints/CCalls and Add/Sub on Int64
values.  The former would, at the moment, complicate the generic paths
that simultaneously iterate over B3 Values and Air args. The latter
(Add/Sub) is lowered to the pre-existing Add64/Sub64 Air forms, which
avoids the need to track the carry bit.

Since we need to interface with the above values that use Int64s, this
patch also introduces two new B3 opcodes (that are only used on
32-bits):

- TruncHigh: extracts the high Int32 of an Int64 value.
- Stitch: creates an Int64 out of two Int32 values.

They need to be defined on 64-bits too because B3ValueInlines.h
references opcodes in a macro definition.

This removes the previously added lowering code to
B3LowerToAir32_64.cpp. It implements enough ops in B3LowerInt64.cpp to
keep i64-call.js and cc-int-to-int.js working with OMG on ARMv7.

It also fixes a bug in the Air forms for Add64/Sub64.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3Generate.cpp:
(JSC::B3::generateToAir):
* Source/JavaScriptCore/b3/B3LowerInt64.cpp: Added.
(JSC::B3::lowerInt64):
* Source/JavaScriptCore/b3/B3LowerInt64.h: Added.
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3StackmapSpecial.cpp:
(JSC::B3::StackmapSpecial::isArgValidForRep):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
(JSC::B3::Value::typeFor):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueRep.h:
(JSC::B3::ValueRep::ValueRep):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:

Canonical link: https://commits.webkit.org/280764@main
